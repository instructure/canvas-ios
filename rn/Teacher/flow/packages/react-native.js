//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
