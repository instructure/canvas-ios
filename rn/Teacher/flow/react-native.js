//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
