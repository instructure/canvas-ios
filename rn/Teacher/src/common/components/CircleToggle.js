//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { PureComponent } from 'react'
import {
  TouchableOpacity,
  StyleSheet,
  View,
} from 'react-native'
import branding from '../branding'
import { Text } from '../text'

type Props = {
  on: boolean,
  children?: any,
  style?: any,
  itemID?: string,
  value: any,
  onPress: Function,
  onLongPress?: (itemID: string, buttonFrame: { x: number, y: number, width: number, height: number }) => void,
  accessibilityLabel?: string,
}

export default class CircleToggle extends PureComponent<Props> {
  buttonViewRef: any

  onPress = () => {
    this.props.onPress(this.props.value, this.props.itemID)
  }

  onLongPress = () => {
    if (this.buttonViewRef == null) {
      return
    }

    this.buttonViewRef.measure((vx: number, vy: number, width: number, height: number, x: number, y: number) => {
      if (this.props.itemID && this.props.onLongPress) {
        this.props.onLongPress(this.props.itemID, { x, y, width, height })
      }
    })
  }

  getRef = (ref: any) => {
    this.buttonViewRef = ref
  }

  render () {
    let viewStyle = [circleButtonStyles.container, this.props.style]
    let textStyle = {
      fontSize: 20,
      fontWeight: '500',
      color: '#8B969E',
    }
    if (this.props.on) {
      viewStyle.push({
        backgroundColor: branding.primaryBrandColor,
        borderWidth: 0,
      })
      textStyle.color = 'white'
    }

    // don't set the longPress if the consumer didn't provide a handler
    const longPress = this.props.onLongPress
      ? this.onLongPress
      : undefined

    let traits = ['button']
    if (this.props.on) {
      traits.push('selected')
    }

    return (
      <TouchableOpacity
        {...this.props}
        accessible
        accessibilityTraits={traits}
        onPress={this.onPress}
        onLongPress={longPress}
      >
        <View style={viewStyle} ref={this.getRef}>
          {typeof this.props.children === 'object'
            ? this.props.children
            : <Text style={textStyle} accessible={false}>{this.props.children}</Text>
          }
        </View>
      </TouchableOpacity>
    )
  }
}

const circleButtonStyles = StyleSheet.create({
  container: {
    borderColor: '#C7CDD1',
    borderWidth: StyleSheet.hairlineWidth,
    minWidth: 48,
    height: 48,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 24,
    flex: 1,
    paddingHorizontal: 8,
  },
})
