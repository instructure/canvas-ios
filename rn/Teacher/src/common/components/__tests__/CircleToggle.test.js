//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React from 'react'
import { Image } from 'react-native'
import CircleToggle from '../CircleToggle'
import Images from '../../../images'
import explore from '../../../../test/helpers/explore'
import renderer from 'react-test-renderer'

jest.mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'Button')

let circleToggleProps = {
  on: false,
  onPress: jest.fn(),
  value: '1',
  itemID: '1',
  testID: 'circle-button',
}

test('renders circle toggle correctly', () => {
  let tree = renderer.create(
    <CircleToggle {...circleToggleProps}>4</CircleToggle>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders circle toggle when on correctly', () => {
  let tree = renderer.create(
    <CircleToggle {...circleToggleProps} on={true}>4</CircleToggle>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders circle toggle with an image', () => {
  let tree = renderer.create(
    <CircleToggle {...circleToggleProps}><Image source={Images.add}/></CircleToggle>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('calls onPress with the value', () => {
  let tree = renderer.create(
    <CircleToggle {...circleToggleProps}>1</CircleToggle>
  ).toJSON()

  let button = explore(tree).selectByID('circle-button') || {}
  button.props.onPress()

  expect(circleToggleProps.onPress).toHaveBeenCalledWith(circleToggleProps.value, circleToggleProps.itemID)
})

test('CircleToggle measures itself on longpress and calls props.onLongPress', () => {
  const onLongPress = jest.fn()
  let button = renderer.create(
    <CircleToggle {...circleToggleProps} onLongPress={onLongPress}>
      12
    </CircleToggle>
  )

  type MeasureFunc = (vx: number, vy: number, width: number, height: number, x: number, y: number) => void
  const measure = jest.fn((callback: MeasureFunc) => {
    callback(1, 2, 3, 4, 5, 6)
  })
  button.getInstance().buttonViewRef = { measure }

  const toggle = explore(button.toJSON()).selectByID(circleToggleProps.testID)
  toggle && toggle.props.onLongPress()
  expect(measure).toHaveBeenCalled()
  expect(onLongPress).toHaveBeenCalled()
})
