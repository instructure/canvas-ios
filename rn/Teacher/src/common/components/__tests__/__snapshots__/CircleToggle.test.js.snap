// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders circle toggle correctly 1`] = `
<Button
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  accessible={true}
  itemID="1"
  on={false}
  onPress={[Function]}
  testID="circle-button"
  value="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#C7CDD1",
          "borderRadius": 24,
          "borderWidth": 0.5,
          "flex": 1,
          "height": 48,
          "justifyContent": "center",
          "minWidth": 48,
          "paddingHorizontal": 8,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={false}
      style={
        Array [
          Object {
            "color": "#2D3B45",
            "fontFamily": "Lato-Regular",
            "fontSize": 16,
          },
          Object {
            "color": "#556572",
            "fontSize": 20,
            "fontWeight": "500",
          },
        ]
      }
    >
      4
    </Text>
  </View>
</Button>
`;

exports[`renders circle toggle when on correctly 1`] = `
<Button
  accessibilityTraits={
    Array [
      "button",
      "selected",
    ]
  }
  accessible={true}
  itemID="1"
  on={true}
  onPress={[Function]}
  testID="circle-button"
  value="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#C7CDD1",
          "borderRadius": 24,
          "borderWidth": 0.5,
          "flex": 1,
          "height": 48,
          "justifyContent": "center",
          "minWidth": 48,
          "paddingHorizontal": 8,
        },
        undefined,
        Object {
          "backgroundColor": "#008EE2",
          "borderWidth": 0,
        },
      ]
    }
  >
    <Text
      accessible={false}
      style={
        Array [
          Object {
            "color": "#2D3B45",
            "fontFamily": "Lato-Regular",
            "fontSize": 16,
          },
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
            "fontWeight": "500",
          },
        ]
      }
    >
      4
    </Text>
  </View>
</Button>
`;

exports[`renders circle toggle with an image 1`] = `
<Button
  accessibilityTraits={
    Array [
      "button",
    ]
  }
  accessible={true}
  itemID="1"
  on={false}
  onPress={[Function]}
  testID="circle-button"
  value="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#C7CDD1",
          "borderRadius": 24,
          "borderWidth": 0.5,
          "flex": 1,
          "height": 48,
          "justifyContent": "center",
          "minWidth": 48,
          "paddingHorizontal": 8,
        },
        undefined,
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/images/Add.png",
        }
      }
    />
  </View>
</Button>
`;
